akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraOffsetStoreImpl.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.projection.cassandra.internal
+
+import java.util.UUID
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.persistence.query
+import akka.projection.ProjectionId
+import akka.projection.cassandra.CassandraOffsetStore
+import akka.stream.alpakka.cassandra.scaladsl.CassandraSession
+import com.datastax.oss.driver.api.core.cql.Row
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object CassandraOffsetStoreImpl {
+  private val StringManifest = "STR"
+  private val LongManifest = "LNG"
+  private val IntManifest = "INT"
+  private val SequenceManifest = "SEQ"
+  private val TimeBasedUUIDManifest = "TBU"
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class CassandraOffsetStoreImpl(session: CassandraSession)(implicit ec: ExecutionContext)
+    extends CassandraOffsetStore {
+  import CassandraOffsetStoreImpl._
+
+  // FIXME make keyspace and table names configurable
+  val keyspace = "akka_projection"
+  val table = "offset_store"
+
+  override def readOffset[Offset](projectionId: ProjectionId): Future[Option[Offset]] = {
+    session
+      .selectOne(s"SELECT offset, manifest FROM $keyspace.$table WHERE projection_id = ?", projectionId.id)
+      .map(fromRowToOffset)
+  }
+
+  def saveOffset[Offset](projectionId: ProjectionId, offset: Offset): Future[Done] = {
+    val (offsetStr, manifest) = stringOffsetAndManifest(offset)
+    session.executeWrite(
+      s"INSERT INTO $keyspace.$table (projection_id, offset, manifest) VALUES (?, ?, ?)",
+      projectionId.id,
+      offsetStr,
+      manifest)
+  }
+
+  /**
+   * Deserialize an offset from a result row.
+   * The offset is converted from its string representation to its real type.
+   */
+  private def fromRowToOffset[Offset](maybeRow: Option[Row]): Option[Offset] = {
+    maybeRow match {
+      case Some(row) =>
+        val offsetStr = row.getString("offset")
+        val manifest = row.getString("manifest")
+        val offset = (manifest match {
+          case StringManifest        => offsetStr
+          case LongManifest          => offsetStr.toLong
+          case IntManifest           => offsetStr.toInt
+          case SequenceManifest      => query.Offset.sequence(offsetStr.toLong)
+          case TimeBasedUUIDManifest => query.Offset.timeBasedUUID(UUID.fromString(offsetStr))
+        }).asInstanceOf[Offset]
+        Some(offset)
+      case None =>
+        None
+    }
+  }
```

Comment:
Why is the offsetStore responsible to handle the offset types/manifests?

Could `readOffset` have an extra argument `offsetDeser: (ByteString, String) => Offset` so the it's not responsibility of the offset store to manage deserialization.

Reply:
Captured that question in https://github.com/akka/akka-projection/issues/60
Let's discuss there.